### PR TITLE
Moved hash into backend. Closes #515

### LINKED ISSUE #515: Move BackendCall.hash to JobDescriptor

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

**Comment by cjllanwarne** [comment]
@geoffjentry @kshakir please!

**Comment by kshakir** [comment]
:+1:

**Comment by Horneth** [comment]
:+1:

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala [hunk 1]
```diff
@@ -166,4 +166,53 @@ trait Backend {
...
       descriptor,
       Option(descriptor.workflowDescriptor.workflowOptions))
+
+  /** Given the specified value for the Docker hash, return the overall hash for this `BackendCall`. */
+  private def hashGivenDockerHash(jobDescriptor: BackendCallJobDescriptor)(dockerHash: Option[String]): ExecutionHash = {
+    val call = jobDescriptor.key.scope
+    val runtimeAttributes = jobDescriptor.callRuntimeAttributes
+    val orderedInputs = jobDescriptor.locallyQualifiedInputs.toSeq.sortBy(_._1)
+    val orderedOutputs = call.task.outputs.sortWith((l, r) => l.name > r.name)
+    val orderedRuntime = Seq(
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: You're creating a List, if you want a List just call it a List

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala [hunk 2]
```diff
@@ -166,4 +166,53 @@ trait Backend {
...
+      orderedInputs map { case (k, v) => s"$k=${v.computeHash(jobDescriptor.workflowDescriptor.fileHasher).value}" } mkString "\n",
+      orderedRuntime map { case (k, v) => s"$k=$v" } mkString "\n",
+      orderedOutputs map { o => s"${o.wdlType.toWdlString} ${o.name} = ${o.requiredExpression.toWdlString}" } mkString "\n"
+    ).mkString("\n---\n").md5Sum
+
+    ExecutionHash(overallHash, dockerHash)
+  }
+
+  /**
+    * Compute a hash that uniquely identifies this call
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: Is `call` the right term to be using now?

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: I think it probably is, looking at this more but still a question

**Comment by geoffjentry** [comment]
:+1:

**Pull request merged**